### DISCUSSION TITLE
Why pad inlineCode's value with spaces?

### DISCUSSION THREAD

**Discussion by vvenv** [syntax-tree/unist#46] [Q&A] [answered]
No idea of these:

```js
  // If this is not just spaces or eols (tabs don’t count), and either the
  // first or last character are a space, eol, or tick, then pad with spaces.
  if (
    /[^ \r\n]/.test(value) &&
    (/[ \r\n`]/.test(value.charAt(0)) ||
      /[ \r\n`]/.test(value.charAt(value.length - 1)))
  ) {
    value = ' ' + value + ' '
  }
```
https://github.com/syntax-tree/mdast-util-to-markdown/blob/main/lib/handle/inline-code.js#L28-L34

A inline-code with value ` inline code` (space+inline+space+code) will transform to `  inline code ` (space+space+inline+space+code+space), with unexpected extra spaces.

**Comment by wooorm** [comment, marked as answer]
This behavior is needed to roundtrip the value (see the comment above it): https://spec.commonmark.org/0.30/#code-span.
The padding has to be done on both sides: it doesn’t work if it’s only on one side, then it changes the code.

As you can see from your example, they are equivalent: the spaces around it are dropped.

See some example: https://spec.commonmark.org/dingus/?text=1.%20%60%20%60%20one%20space%20(persists)%0A1.%20%60%20%20%20%20%60%20four%20spaces%20(persists)%0A1.%20%60a%20%60%20final%20space%20(persists)%0A1.%20%60%20a%60%20initial%20space%20(persists)%0A1.%20%60%20a%20%60%20initial%20and%20final%20space%20(dropped)%0A1.%20%60%60%20%60a%60%20%60%60%20spaces%20around%20ticks%20(dropped)%0A1.%20%60%20%60%60a%60%60%20%60%20spaces%20around%20ticks%2C%20inverted%20(dropped)%0A

**Reply by vvenv** [reply]
> This behavior is needed to roundtrip the value (see the comment above it): https://spec.commonmark.org/0.30/#code-span.
> The padding has to be done on both sides: it doesn’t work if it’s only on one side, then it changes the code.
> 
> As you can see from your example, they are equivalent: the spaces around it are dropped.
> 
> See some example: https://spec.commonmark.org/dingus/?text=1.%20%60%20%60%20one%20space%20(persists)%0A1.%20%60%20%20%20%20%60%20four%20spaces%20(persists)%0A1.%20%60a%20%60%20final%20space%20(persists)%0A1.%20%60%20a%60%20initial%20space%20(persists)%0A1.%20%60%20a%20%60%20initial%20and%20final%20space%20(dropped)%0A1.%20%60%60%20%60a%60%20%60%60%20spaces%20around%20ticks%20(dropped)%0A1.%20%60%20%60%60a%60%60%20%60%20spaces%20around%20ticks%2C%20inverted%20(dropped)%0A

As edited above, (space+inline+space+code) will transform to (space+space+inline+space+code+space).

https://spec.commonmark.org/dingus/?text=%60%20inline%20code%60%0A&smart=1

I make a new PR to mdast-util-to-markdown: https://github.com/syntax-tree/mdast-util-to-markdown/pull/33

**Reply by wooorm** [reply]
Yes, this happens. But that’s okay! The added spaces are ignored by markdown parsers.
However, your proposed behavior: adding only an initial or final space, **breaks**. It removes a space that was in the code. It breaks roundtripping.

**Reply by vvenv** [reply]
If make less changes of original data, we can make this happen:

```js
const markdown = ` inline code`;
const html = markdownToHtml(markdown);

assert(htmlToMarkdown(html) === markdown)
```

It's usefully for a WYSIWYG Markdown editor.

**Reply by wooorm** [reply]
It sounds like that is a bug in your md->html->md pipeline.

Otherwise, no, we can’t change it, doing so changes the value of the code.

Try roundtripping with both mdast-util-from-markdown and mdast-util-to-markdown to see that it changes the AST.

**Reply by wooorm** [reply]
We might be able to leave it as-is if there’s a `[ \r\n]` on *one* side but not both, perhaps?
It’s definitely needed if there’s a space on *both* sides or a tick on one or both sides 🤔

**Reply by wooorm** [reply]
Like this?

```js
if (
  /[^ \r\n]/.test(value) &&
  ((/^[ \r\n]/.test(value) && /[ \r\n]$/.test(value)) ||
    /^`|`$/.test(value))
) {
  // Pad
}
```

**Reply by vvenv** [reply]
https://github.com/commonmark/commonmark-spec-web/blob/gh-pages/js/commonmark.js#L7824-L7833

```js
if (
    contents.length > 0 &&
    contents.match(/[^ ]/) !== null &&
    contents[0] == " " &&
    contents[contents.length - 1] == " "
) {
    node._literal = contents.slice(1, contents.length - 1);
} else {
    node._literal = contents;
}
```

There is no whitespace *padding* in spec, just *dropping* and *persisting*

**Reply by wooorm** [reply]
Indeed, parsing drops or persists. So micromark/mdast-util-to-markdown do that.
Because `mdast-util-to-markdown` *serializes*, we need to prevent important information from being removed:

```js
import assert from 'node:assert'
import {fromMarkdown} from 'mdast-util-from-markdown'
import {toMarkdown} from 'mdast-util-to-markdown'
import {removePosition} from 'unist-util-remove-position'

const input = 'a `  b  ` c'

const treeA = removePosition(fromMarkdown(input))
const treeB = removePosition(fromMarkdown(toMarkdown(treeA)))

assert.deepStrictEqual(treeA, treeB)
console.dir(treeA, {depth: null})
```
